pandas/io/excel.py

```diff
@@ -207,21 +209,22 @@ def __init__(self, io, **kwds):
         if engine is not None and engine != 'xlrd':
             raise ValueError("Unknown engine: %s" % engine)
 
-        if isinstance(io, compat.string_types):
-            if _is_s3_url(io):
-                buffer, _, _ = get_filepath_or_buffer(io)
-                self.book = xlrd.open_workbook(file_contents=buffer.read())
-            elif _is_url(io):
-                data = _urlopen(io).read()
-                self.book = xlrd.open_workbook(file_contents=data)
-            else:
-                self.book = xlrd.open_workbook(io)
-        elif engine == 'xlrd' and isinstance(io, xlrd.Book):
+        # If io is a url, want to keep the data as bytes so can't pass
+        # to get_filepath_or_buffer()
+        if _is_url(io):
```

Comment:
I haven't looked closely at the changes in this PR (will followup later), but I'm partway through refactoring our S3 handling to use [s3fs](http://s3fs.readthedocs.org/en/latest/), so I might be changing this soon anyway. My branch is [here](https://github.com/TomAugspurger/pandas/tree/s3fs).